### ISSUE TITLE
Maintaining state through render passes  

### ISSUE THREAD

**Issue by ev1stensberg** [facebook/react#6543]
Someone interested in trying a "fix" on this? Meaning you would still have a reference(key, ish) to state through a render pass, but it will be destructed and you would only have the reference left on destruction. On re-render, the state rebuilds the previous one, kinda like hot reloading, but it watches how the previous state rendered and tries to re-render out of that, and then builds the changes made in the new render. Dunno how this will be in-action though.

**Is this something worth chasing?**

**Comment by ev1stensberg** [comment]
Basically what I propose is reaching elements in the garbage collector, if that's possible. I'd imagine a plot-twist Mark & Sweep algorithm.

**Comment by ev1stensberg** [comment]
Edit: should we try tweaking ref to do this for us?

**Comment by gaearon** [comment]
I’m not sure I understand what you describe. Can you rephrase this from the use case perspective? What are you trying to accomplish, and why?

**Comment by ev1stensberg** [comment]
I'll put up a use case:

 ``` 
render: function() {
      <TextInput
        ref={this.state}/>
  },
 ```

Then you would have another component using your `<TextInput  />` but ref gets left alone if you pass it through state. Means we should try to destruct it and rebuild it out of a reference, without it having to do a full re-render. Perhaps this somewhat weird caching solution with the garbage.

"_**Refs are a great way to send a message to a particular child instance in a way that would be inconvenient to do via streaming Reactive props and state. They should, however, not be your go-to abstraction for flowing data through your application. By default, use the Reactive data flow and save refs for use cases that are inherently non-reactive.**_"

Means, if I got this correct:

Component ----> render ----> ref component instance ----> Component,

this way we could "force" React to have track of the component now and then.

Not sure how this is healthy for our common friend "memory" though

I'd like this to stay open as a discussion if people get the point. Otherwise -- Close

**Comment by gaearon** [comment]
I don’t understand what `ref={this.state}` is supposed to do. Refs give you access to the component instance in case you want to call methods on it imperatively. There are string refs (e.g. `ref='something') that make the instance available as `this.refs.something`, and callback refs (e.g. `ref={instance => this.something = instance}`) which let you do the same thing but are more generic and composable. I don’t understand what you are trying to do with `ref={this.state}` though.

**Comment by ev1stensberg** [comment]
`ref={this.state}` will have the same functionality as `ref={instance => this.something = instance}` except it will keep track of the state from  now and then. `this.state` would be the state of the component in its simplest form without this functionality being mounted in any way. With this functionality, I'd imagine it would be, somewhat similar to callback refs, but with the state in mind, not the refs.

**Comment by ev1stensberg** [comment]
The example, used refs mainly because it has callbacks by nature by the way.

**Comment by jimfb** [comment]
I don't think we'd do this with `ref={this.state}`, and we certainly can't access objects that have been GC'd (as of ES2015).  The general issue of externalizing state is being tracked as https://github.com/facebook/react/issues/4595

**Comment by ev1stensberg** [comment]
Reading the discussion now, thanks for the heads up! This was actually somewhat similar as I was looking ofr @jimfb 😄
